File: 17_模式匹配/isSubarray.go
package main

import "fmt"

func main() {
	fmt.Println(IsSubstring("abc", "ab"))
}

// O(n+m)判断`shorter`是否是`longer`的子串.
func IsSubstring(longer, shorter string) bool {
	if len(shorter) > len(longer) {
		return false
	}
	if len(shorter) == 0 {
		return true
	}
	ords1 := make([]int, len(longer))
	for i, c := range longer {
		ords1[i] = int(c)
	}
	ords2 := make([]int, len(shorter))
	for i, c := range shorter {
		ords2[i] = int(c)
	}
	return IsSubarray(ords1, ords2)
}

// O(n+m)判断`shorter`是否是`longer`的子数组.
func IsSubarray(longer, shorter []int) bool {
	if len(shorter) > len(longer) {
		return false
	}
	if len(shorter) == 0 {
		return true
	}
	n, m := len(longer), len(shorter)
	st := make([]int, 0, n+m)
	st = append(st, shorter...)
	st = append(st, longer...)
	z := zAlgo(st)
	for i := m; i < n+m; i++ {
		if z[i] >= m {
			return true
		}
	}
	return false
}

func zAlgo(seq []int) []int {
	n := len(seq)
	if n == 0 {
		return nil
	}
	z := make([]int, n)
	j := 0
	for i := 1; i < n; i++ {
		var k int
		if j+z[j] <= i {
			k = 0
		} else {
			k = min(j+z[j]-i, z[i-j])
		}
		for i+k < n && seq[k] == seq[i+k] {
			k++
		}
		if j+z[j] < i+z[i] {
			j = i
		}
		z[i] = k
	}
	z[0] = n
	return z
}

func min(a, b int) int {
	if a <= b {
		return a
	}
	return b
}

File: 17_模式匹配/isSubarray.ts
/**
 * O(n+m) 判断`shorter`是否是`longer`的子串.
 */
function isSubarray<T extends ArrayLike<unknown>>(longer: T, shorter: T): boolean {
  if (shorter.length > longer.length) return false
  if (!shorter.length) return true
  const n = longer.length
  const m = shorter.length
  const st = Array(m + n)
  for (let i = 0; i < m; i++) st[i] = shorter[i]
  for (let i = 0; i < n; i++) st[i + m] = longer[i]
  const z = zAlgo(st)
  for (let i = m; i < n + m; i++) {
    if (z[i] >= m) return true
  }
  return false
}

function zAlgo(seq: ArrayLike<unknown>): Uint32Array {
  const n = seq.length
  if (!n) return new Uint32Array(0)
  const z = new Uint32Array(n)
  let j = 0
  for (let i = 1; i < n; i++) {
    let k = 0
    if (j + z[j] > i) {
      k = Math.min(j + z[j] - i, z[i - j])
    }
    while (i + k < n && seq[k] === seq[i + k]) k++
    if (j + z[j] < i + z[i]) j = i
    z[i] = k
  }
  z[0] = n
  return z
}

export { isSubarray }

if (require.main === module) {
  console.log(isSubarray('abc', 'ab'))
}
